src/algebra/module/basic.lean

```diff
@@ -461,19 +455,64 @@ end
 
 end add_monoid_hom
 
-section module_division_ring
-/-! Some tests for the vanishing of elements in modules over division rings. -/
+section no_zero_smul_divisors
+/-! ### `no_zero_smul_divisors`
+
+This section defines the `no_zero_smul_divisors` class, and includes some tests
+for the vanishing of elements (especially in modules over division rings).
+-/
+
+/-- `no_zero_smul_divisors R M` states that a scalar multiple is `0` only if either argument is `0`.
+
+The main application of `no_zero_smul_divisors R M`, when `M` is a semimodule,
+is the result `smul_eq_zero`: a scalar multiple is `0` iff either argument is `0`.
+
+It is a generalization of the `no_zero_divisors` class to heterogeneous multiplication.
+-/
+class no_zero_smul_divisors (R M : Type*) [has_zero R] [has_zero M] [has_scalar R M] : Prop :=
+(eq_zero_or_eq_zero_of_smul_eq_zero : ∀ {c : R} {x : M}, c • x = 0 → c = 0 ∨ x = 0)
+
+export no_zero_smul_divisors (eq_zero_or_eq_zero_of_smul_eq_zero)
+
+section semimodule
+
+variables [semiring R] [add_comm_monoid M] [semimodule R M]
+
+instance no_zero_smul_divisors.of_no_zero_divisors [no_zero_divisors R] :
+  no_zero_smul_divisors R R :=
+⟨λ c x, no_zero_divisors.eq_zero_or_eq_zero_of_mul_eq_zero⟩
```

Comment:
This only needs `monoid R` I think, so can probably move to `group_action` too

Reply:
A `monoid_with_zero` has no `smul`, the `has_scalar R R` here comes from `semiring.to_semimodule`. No objections to moving that instance lower down the hierarchy, but I don't think it's appropriate for this PR.

Reply:
It does in the presence of `mul_action.regular`

Reply:
To be clear, I'm not asking you to generalize further in this PR, it looks great as is!